gradio/analytics.py

```diff
@@ -29,60 +31,91 @@ async def pyodide_pyfetch(*_args, **_kwargs):
+    import gradio as gr
+
+    classes_to_check = (
+        gr.components.Component.__subclasses__()
+        + gr.blocks.BlockContext.__subclasses__()  # type: ignore
+    )
+    subclasses = []
+
+    while classes_to_check:
+        subclass = classes_to_check.pop()
+        children = subclass.__subclasses__()
+        if children:
+            classes_to_check.extend(children)
+        if subclass.__module__.startswith("gradio."):
+            subclasses.append(get_block_name(subclass))
+
+    return subclasses
+
+
 def analytics_enabled() -> bool:
     """
     Returns: True if analytics are enabled, False otherwise.
     """
     return os.getenv("GRADIO_ANALYTICS_ENABLED", "True") == "True"
 
 
-def _do_analytics_request(url: str, data: dict[str, Any]) -> None:
+def _do_analytics_request(topic: str, data: dict[str, Any]) -> None:
     if wasm_utils.IS_WASM:
         asyncio.ensure_future(
             _do_wasm_analytics_request(
-                url=url,
+                topic=topic,
                 data=data,
             )
         )
     else:
         threading.Thread(
             target=_do_normal_analytics_request,
             kwargs={
-                "url": url,
+                "topic": topic,
                 "data": data,
             },
         ).start()
 
 
-def _do_normal_analytics_request(url: str, data: dict[str, Any]) -> None:
+def _do_normal_analytics_request(topic: str, data: dict[str, Any]) -> None:
     data["ip_address"] = get_local_ip_address()
     try:
-        httpx.post(url, data=data, timeout=5)
-    except (httpx.ConnectError, httpx.ReadTimeout):
-        pass  # do not push analytics if no network
+        _send_telemetry_in_thread(
+            topic=topic,
+            library_name="gradio",
+            library_version=data.get("version"),
+            user_agent=data,
+        )
+    except Exception:
+        pass
 
 
-async def _do_wasm_analytics_request(url: str, data: dict[str, Any]) -> None:
+async def _do_wasm_analytics_request(topic: str, data: dict[str, Any]) -> None:
     data["ip_address"] = await get_local_ip_address_wasm()
-
-    # We use urllib.parse.urlencode to encode the data as a form.
-    # Ref: https://docs.python.org/3/library/urllib.request.html#urllib-examples
-    body = urllib.parse.urlencode(data).encode("ascii")
-    headers = {
-        "Content-Type": "application/x-www-form-urlencoded",
-    }
-
     try:
+        path = "/".join(quote(part) for part in topic.split("/") if len(part) > 0)
+        url = f"https://huggingface.co/api/telemetry/{path}"
+        headers = build_hf_headers(
+            token=False,  # no need to send a token for telemetry
+            library_name="gradio",
+            library_version=data.get("version"),
+            user_agent=data,
+        )
         await asyncio.wait_for(
-            pyodide_pyfetch(url, method="POST", headers=headers, body=body),
+            pyodide_pyfetch(url, method="HEAD", headers=headers),
```

Comment:
I originally wrote this to use `_send_telemetry_in_thread` but the analytics requests were not showing up in the webworker network tab.

If I use `pyodide_pyfetch`, they show up in the network tab. Still not sure if this works 100% though.

So I think we need this lower-level approach for wasm and can't rely on higher level functions in hub python library. 


<img width="531" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/7aff0657-6dc6-457d-808f-5c0fc4003d01">

Reply:
If `_send_telemetry_in_thread` uses threading as its name implies, it [doesn't work](https://pyodide.org/en/stable/usage/wasm-constraints.html#included-but-not-working-modules) on Lite/Pyodide.

I think using `pyodide_pyfetch` is a right approach on Lite and it should work.

Reply:
I reverted the changes for wasm since we can't set the user-agent on chrome!

Reply:
Ref: https://stackoverflow.com/a/42815264/13103190